homeassistant/components/remootio/config_flow.py

```diff
@@ -0,0 +1,172 @@
+        ),
+    },
+    extra=REMOVE_EXTRA,
+)
+
+DEVICE_NAME = "Remootio Device"
+
+
+async def validate_input(hass: HomeAssistant, data: dict[str, Any]) -> dict[str, Any]:
+    """Validate the user input."""
+
+    _LOGGER.debug("Validating input... Input [%s]", data)
+
+    data = INPUT_VALIDATION_SCHEMA(data)
+
+    connection_options: ConnectionOptions = ConnectionOptions(
+        data[CONF_HOST], data[CONF_API_SECRET_KEY], data[CONF_API_AUTH_KEY]
+    )
+
+    device_serial_number: str = await get_serial_number(
+        hass, connection_options, _LOGGER
+    )
+
+    return {
+        CONF_TITLE: f"{DEVICE_NAME} (Host: {data[CONF_HOST]}, S/N: {device_serial_number})",
+        CONF_SERIAL_NUMBER: device_serial_number,
+    }
+
+
+class ConfigFlow(config_entries.ConfigFlow, domain=DOMAIN):
+    """Handle a config flow for Remootio."""
+
+    VERSION = 1
+
+    def __init__(self) -> None:
+        """Initialize the class."""
+
+    async def async_step_user(
+        self, user_input: dict[str, Any] | None = None
+    ) -> FlowResult:
+        """Handle the user step."""
+        user_input = user_input or {}
+        errors = {}
+
+        if len(user_input) != 0:
+            try:
+                info = await validate_input(self.hass, user_input)
+            except UnsupportedRemootioDeviceError:
+                _LOGGER.debug("Remootio device isn't supported", exc_info=True)
+                return self.async_abort(reason="unsupported_device")
+            except vol.MultipleInvalid as ex:
+                _LOGGER.error(
+                    "Invalid user input. MultipleInvalid.Errors [%s]", ex.errors
+                )
+                for error in ex.errors:
+                    _LOGGER.debug(
+                        "Error [%s] Path [%s]", error.__class__.__name__, error.path[0]
+                    )
+                    if isinstance(error, RequiredFieldInvalid):
+                        errors[str(error.path[0])] = f"{error.path[0]}_required"
+                    else:
+                        errors[str(error.path[0])] = f"{error.path[0]}_invalid"
+            except RemootioClientConnectionEstablishmentError:
+                _LOGGER.error("Can't connect to Remootio device")
+                errors["base"] = "cannot_connect"
+            except RemootioClientAuthenticationError:
+                _LOGGER.error("Can't authenticate by the Remootio device")
+                errors["base"] = "invalid_auth"
+            except Exception:  # pylint: disable=broad-except
+                _LOGGER.exception("Unexpected exception/error")
+                errors["base"] = "unknown"
+            else:
+                await self.async_set_unique_id(info[CONF_SERIAL_NUMBER])
+                self._abort_if_unique_id_configured(user_input)
+
+                user_input[CONF_SERIAL_NUMBER] = info[CONF_SERIAL_NUMBER]
+
+                return self.async_create_entry(title=info[CONF_TITLE], data=user_input)
```

Comment:
You are still storing the non schema validated data here. So data is in config entry is not uppercased like you validation schema does. It's not important. But just thought I'd mention.

Reply:
Adapted.